### PR TITLE
[Frontend] Defer runtime imports until CLI dispatch

### PR DESCRIPTION
**Pull request by matteso1**
### Prior art

#41518 addresses the same CLI feedback problem and additionally accelerates misspelled and ill-typed arguments. This PR is the consolidation point for the core vLLM import-light parser path requested in review.

I coordinated the overlap in #41518 before opening this PR. Substantive import-light work from @netanel-haber is credited in the commit history and adapted here with the same goal: one canonical parser, not a duplicate compact schema. This PR does not claim #41518's broader invalid-argument surface.

This also retains the generic CLI split that @simon-mo asked to separate from #51360: https://github.com/vllm-project/vllm/pull/51360#discussion_r3815210367

### The bug

The `vllm` console entry point imports the runtime graph before it knows which command the user selected. Help rendering therefore waits for Torch, platform, executor, and worker imports that it does not use.

### The fix

- Use the canonical serve parser for plain, full, section, and flag help.
- Carry the import-light boundary through `EngineArgs` and the config tree.
- Make `vllm.env_override` the single owner of pre-Torch settings and targeted post-import patches; CLI code retains only its separate spawn policy.
- Lazy-load runtime command, platform, OpenAI, tool-parser, and config dependencies at their use sites.
- Remove the duplicate compact serve parser and distributed patch call sites.
- Add the `cli-import-boundary` agent skill (`.agents/skills/`, mirrored from `.claude/skills/`) so future CLI changes keep the import-light contract and know its enforcing tests.

The callback registry wraps only exact `PathFinder` targets, restores original loaders before applying patches, retains registrations when target or callback execution fails, and unregisters only after all callbacks succeed.

### Scope and compatibility

This is a stock-vLLM CLI/help change. It does not add an external plugin metadata API or promise import-light behavior for arbitrary third-party plugin code. No model execution or output behavior is intentionally changed.

Because the help path no longer runs `load_general_plugins()` or `current_platform.pre_register_and_update(parser)`, serve help renders the stock option set: choices registered by an out-of-tree platform or quantization plugin appear at runtime but not in help. Similarly, `--mm-processor-device` help shows the in-tree device superset with a note that the accepted set is platform dependent, where main rendered the platform-derived subset.

The candidate and current main `c7b0467c25` expose an identical long-option inventory in `vllm serve --help=all`. Plain, full, and section help avoid importing Torch or a concrete platform implementation. Runtime-dependent nested config object dumps are omitted from help rather than constructing platform state; normal parser defaults are unchanged.

Where the import-light boundary reaches a Pydantic config field, the runtime type alias resolves to `object` outside `TYPE_CHECKING`. Each such field (`DeviceConfig.device`, `LoRAConfig.lora_dtype`, `ModelConfig.dtype`) is unioned with a `Literal`, and `arg_utils` resolves `Literal` before any other branch, so option generation is unchanged. `DeviceConfig.compute_hash` contributes no factors, so the compilation cache key is unaffected.

### Measured

Fresh-process medians from n=5 recorded runs per arm and command, after one unrecorded warm-up, on Apple M5 Pro arm64 (18 logical CPUs), macOS 26.5.2, Python 3.13.5. Both source worktrees used the same interpreter, metadata shim, `VLLM_LOGGING_LEVEL=ERROR`, and `PYTHONWARNINGS=ignore`.

| Command | Main `833fa6233c` | Candidate `a570c76c47` | Speedup |
| --- | ---: | ---: | ---: |
| `vllm --help` | 3255.946 ms | 51.796 ms | 62.9x |
| `vllm serve --help` | 3311.459 ms | 213.492 ms | 15.5x |
| `vllm serve --help=all` | 3369.555 ms | 226.539 ms | 14.9x |
| `vllm serve --help=ModelConfig` | 3405.404 ms | 217.980 ms | 15.6x |

These timings were recorded at main `833fa6233c` and candidate `a570c76c47`. The branch has since been rebased onto `c7b0467c25`; the resulting diff is byte-identical (37 files, +1369/-570) and the timings were not re-recorded at the new base.

All 40 recorded calls returned 0 with empty stderr. Top-level, plain serve, and section-help output sizes match current main. Full help retains the same option inventory; its shorter output omits runtime-dependent object dumps.

These measurements cover CLI feedback, not server startup.

### Test plan

Run at candidate `fbbc1fd7e1` on base `c7b0467c25`.

```text
.venv/bin/python -m pytest -q \
  tests/test_env_override_lifecycle.py \
  tests/entrypoints/launchers/test_cli_imports.py \
  tests/entrypoints/launchers/test_launch_cli.py \
  tests/engine/test_arg_utils.py \
  tests/quantization/test_quantization_config_args.py \
  tests/quantization/test_fp8_per_channel.py \
  tests/config/test_multimodal_config.py \
  tests/cuda/test_cuda_compatibility_path.py \
  tests/compile/test_inductor_fallback_allow_list_patch.py \
  tests/test_fxgraphcache_pickle_patch.py
# 240 passed, 2 skipped

.venv/bin/python -m pytest -q tests/v1/metrics/test_perf_metrics.py
# 56 passed

.venv/bin/python -m pytest -q tests/test_config.py \
  -k 'revision_not_resolved_when_weights_differ_from_model or revision_resolved_when_weights_match_model or dsa_models_default_to_mrv2_and_breakable_cudagraph'
# 8 passed, 221 deselected

uvx pre-commit run --from-ref c7b0467c25 --to-ref fbbc1fd7e1
# all applicable hooks passed, including Python 3.10 mypy and root lazy-import checks
```

Follow-up `bc6bec71c3` gives both new suites a Buildkite collection point: it moves `tests/entrypoints/test_cli_imports.py` into `tests/entrypoints/launchers/`, which the Entrypoints Unit job runs, and adds `tests/test_env_override_lifecycle.py` to the Engine Core job with `vllm/env_override.py` as a source dependency. Follow-up `8d33965128` adds the `cli-import-boundary` agent skill, and `0e6378763a` drops an unused helper.

A review pass then produced `308ef59976` (patch registry delegates to later `sys.meta_path` finders so editable-install targets still get patched, mid-import registration warns instead of raising, torch version lookup falls back to `find_spec` when distribution metadata is absent, and the CUDA-version test now compares against the installed torch), `c23f564d31` (the CLI import-boundary test also runs from the Engine Core job, whose source dependencies cover the files that can actually break the contract), and `bb0cc21128` (serve help no longer advertises a `tpu` processor device no in-tree platform provides, and `Platform.__getattr__` defers its torch import past the dunder guard).

That new CUDA-version oracle failed on its first CI run (`assert None == '13.0'` on the H200 image) and exposed a real reader bug: torch/version.py declares `cuda` with an annotation, which the `ast.Assign` match skipped. `a759ae5ec2` handles annotated assignments and pins the shape with a synthetic-source test; `9ed4b9acc2` keeps the match mypy clean. `a342da04e0` makes the missing-torch negative test block both discovery routes. A hardware gate run on a bare source-tree checkout (no installed vllm metadata) then surfaced a pre-existing CLI fragility: main's `importlib.metadata.version("vllm")` at the `--version` flag kills every CLI invocation with PackageNotFoundError when distribution metadata is absent; `5a39ef13ab` falls back to `vllm.version.__version__`, mirroring the torch version fallback. The diff is now 41 files, +1521/-575.

`tests/test_config.py` requires `huggingface_hub >= 1.28.0` as pinned in `requirements/common.txt`; an older hub in the environment fails at collection on the `ResolvedRevision` import.

No model evaluation was run because this changes CLI import/parser behavior, not model execution or output logic.

AI assistance was used. I take responsibility for the implementation, compatibility boundary, and measurements.

### CODE REVIEW

File: vllm/entrypoints/cli/serve.py [hunk 1]
```diff
@@ -2,41 +2,21 @@
...
-)
-
-logger = init_logger(__name__)
 
-DESCRIPTION = """Launch a local OpenAI-compatible API server to serve LLM
-completions via HTTP. Defaults to Qwen/Qwen3-0.6B if no model is specified.
+if typing.TYPE_CHECKING:
+    from vllm.utils.argparse_utils import FlexibleArgumentParser
+else:
+    FlexibleArgumentParser = argparse.ArgumentParser
```
Comment (hmellor) [inline vllm/entrypoints/cli/serve.py]: This is risky if `FlexibleArgumentParser` is ever instantiated. 


```suggestion
if typing.TYPE_CHECKING:
    from vllm.utils.argparse_utils import FlexibleArgumentParser
```

Let's just make the type hint a forward reference

File: vllm/entrypoints/cli/serve.py
Comment (hmellor) [inline vllm/entrypoints/cli/serve.py]:
```suggestion
    ) -> "FlexibleArgumentParser":
```

File: vllm/entrypoints/cli/serve.py (hunk 1 above)
Comment (matteso1) [reply vllm/entrypoints/cli/serve.py]: Thank you, this is much cleaner. I removed the runtime alias and kept `FlexibleArgumentParser` type-checking-only in `67ee61d01a`.

File: vllm/entrypoints/cli/serve.py
Comment (matteso1) [reply vllm/entrypoints/cli/serve.py]: Fixed the annotation as a forward reference in the same commit. Thank you for the quick review.

File: vllm/__init__.py [hunk 2]
```diff
@@ -66,6 +68,7 @@ def __getattr__(name: str) -> typing.Any:
         from importlib import import_module
 
         if name in MODULE_ATTRS:
+            apply_runtime_environment()
```
Comment (simon-mo) [inline vllm/__init__.py]: is this the right spot for `apply_runtime_environment`?

File: vllm/__init__.py [hunk 3]
```diff
@@ -8,10 +8,12 @@
 
 import typing
 
-# The environment variables override should be imported before any other
-# modules to ensure that the environment variables are set before any
-# other modules are imported.
-import vllm.env_override  # noqa: F401
```
Comment (simon-mo) [inline vllm/__init__.py:14]: this means we are not apply the rest of `env_override` at import any more, the `env_override.py` has another 600 lines of stuff that needed to be ran… perhaps we can just adopt this sprit for the entire env override to make it functional, and apply the override in pre-torch and post-torch phases.

File: vllm/entrypoints/cli/main.py [hunk 4]
```diff
@@ -1,102 +1,155 @@
...
-        spec = find_spec("vllm_omni")
-        if spec is None:
+        if find_spec("vllm_omni") is None:
             logger.error(
                 "--omni flag requires a valid instance of vllm-omni to be installed."
             )
-            sys.exit(1)
+            raise SystemExit(1)
 
+        apply_runtime_environment()
```
Comment (simon-mo) [inline vllm/entrypoints/cli/main.py]: do we need this? i thought we are doing it in `__init__.py`

**Review by simon-mo — COMMENTED** [review]
Overall I think the design is on the right track (delayed import and proper env override).

1. Apply this globally so we do this refactor once
2. Figure out a way to add a test such that if this contract breaks, the test fails. This contract means `vllm serve --help` should not import torch. 
3. Is there a way to make `vllm serve --help=all` to run as fast? 
4. Add a `skill` that compactly instruct new agent when they are modifying the CLI to respect this workload (lazy import, etc); essentially document this for the future code owners.

File: vllm/entrypoints/cli/main.py (hunk 4 above)
Comment (matteso1) [reply vllm/entrypoints/cli/main.py]: yeah you're right, this CLI-owned call is redundant. removing it and keeping environment setup at the package boundary.

File: vllm/__init__.py (hunk 2 above)
Comment (matteso1) [reply vllm/__init__.py]: agreed, this is not the right spot. The lazy attribute hook should not own environment setup, and itll go away with the single-owner rework.

File: vllm/__init__.py (hunk 3 above)
Comment (matteso1) [reply vllm/__init__.py:14]: Done in `447747d766`. env_override.py now works the way you describe: `apply_pre_torch_environment()` owns the pre-Torch settings and runs when vllm is imported, and each Torch-dependent patch is registered as a post-import callback that fires when its exact target module loads (`torch._inductor.config`, `torch._inductor.codegen.wrapper`, `torch._inductor.graph`). If a target is already imported the callback runs immediately, so import order does not change behavior. Nothing in the module waits on Torch at import time anymore, and `tests/test_env_override_lifecycle.py` covers the no-torch import and the callback retry paths.

File: vllm/__init__.py (hunk 2 above)
Comment (matteso1) [reply vllm/__init__.py]: This is gone in `447747d766`. The lazy attribute hook no longer owns any environment setup. `vllm/__init__.py` imports `vllm.env_override`, which applies its own pre-Torch phase, and the CLI keeps only `cli_env_setup()` for the spawn policy.

File: vllm/entrypoints/cli/main.py (hunk 4 above)
Comment (matteso1) [reply vllm/entrypoints/cli/main.py]: Removed in `447747d766`.

**Comment by matteso1** [comment]
Superseded by #54841, which implements the same CLI-feedback outcome at the console-entrypoint/build boundary without changing vLLM runtime or configuration imports. Closing this version so review and CI move to the replacement.
